### ISSUE TITLE
Paths added with -pa and -pz cannot be removed from the module lookup

### ISSUE THREAD

**Issue by josevalim** [erlang/otp#6692]
**Describe the bug**

See the snippet below. I am using Elixir's `ebin` but any `ebin` will work:

```erlang
$ erl -pa lib/elixir/ebin
Erlang/OTP 26 [DEVELOPMENT] [erts-13.1.3] [source-bdf564f8ee] [64-bit] [smp:10:10] [ds:10:10:10] [async-threads:1] [jit]

Eshell V13.1.3 (press Ctrl+G to abort, type help(). for help)
1> elixir:module_info(module).
elixir
2> code:del_path("lib/elixir/ebin").
true
3> elixir:module_info(module).
elixir
4> code:purge(elixir), code:delete(elixir).
true
5> code:purge(elixir), code:delete(elixir).
true
6> elixir:module_info(module).
elixir
```

The reason why this happens is because when trying to load a module, the `code_server` will fallback to the `erl_prim_loader`:

https://github.com/erlang/otp/blob/master/lib/kernel/src/code_server.erl#L1191

And the `erl_prim_loader` always keep it in its code paths all the -pa/-pz entries plus the last call to `path` from `init`.

I would say this is problematic for two reasons:

1. Looking up a module that does not exist can be twice slower because it has to traverse the same paths twice

2. Because some applications use `erl_prim_loader:get_file` to load files, you can end-up in a scenario where we are looking at code in different code paths and getting different results

~I think the solution to this problem may be relatively simple: move the code loading functionality from the `code_server` to `erl_prim_loader`. All of the path functionality in `code_server` (`_path` functions + `dir` + `get_object_code`) feel like they could be extracted rather cleanly. The functionality is also abstracted away via the `code` module.~ (see comments below why this is bad)

**OTP Versions**

Reproduced in Erlang/OTP 25 and master.

**Comment by josevalim** [comment]
I have found more information that perhaps should be considered when working on this issue. It is related to the `-loader inet` mode:

1. Paths added via `code:add_path*` will not use `inet` loading when using `code:ensure_loaded`/`code:get_object_file` and friends. In this case, only the paths given via -pa/-pz or on boot use `inet` lookups

2. However, because `ensure_modules_loaded`/`atomic_load` [retrieves all code paths from the code server and pass them to the `erl_prim_loader`](https://github.com/erlang/otp/blob/91c7d13df72f22f758fb4b41564e91b542107c14/lib/kernel/src/code.erl#L631-L633.), a module loaded via those functions will always use the `inet` mode (which can lead to slightly different behaviour)

I mentioned above that a possible solution is to unify `erl_prim_loader` and `code_server` but that can potentially be a bad idea because doing so would force all module loading to use `inet`, even paths added via `code:add_path*`. In any case, it is clear the duplicate code paths lead to slightly different behaviour here and there.

---

Therefore I would like to propose another possible solution:

1. Keep the code paths separate

2. Once the `code_server` boots, it will call `erl_prim_loader:take_local_paths()`. This function will:
  a. If `efile`, it returns all paths and set `erl_prim_loader` internal paths to `[]`
  b. If `inet`, it splits the local and non-local paths, returning the local ones and keeping the internal `erl_prim_loader` paths as the non-local ones

3. Remove `erl_prim_loader:get_modules/3` and change `atomic_load`/`ensure_loaded_modules` to get those modules from the `code_server`, using the same logic as `ensure_loaded`

This is definitely a smaller change than the one proposed above. It will solve the originally described issue, it will solve the different behaviour found in `atomic_load/ensure_modules_loaded`, and keep the `inet` module working. If that's the road taken, I would suggest documenting that only the boot paths and `-pa`/`-pz` affect `inet`. That's not currently documented.

**Comment by josevalim** [comment]
> Change atomic_load/ensure_loaded_modules to get those modules from the code_server, using the same logic as ensure_loaded

I have implemented this change on PR #6729. This is necessary so both `atomic_load`/`ensure_modules_loaded` can also benefit from code path caching.

**Comment by josevalim** [comment]
Here is a test to be added to `kernel/code_SUITE` that fails.

```erlang
del_path_pa(Config) ->
    PrivDir = proplists:get_value(priv_dir, Config),
    Dir = filename:join(PrivDir, "myebin"),
    filelib:ensure_path(Dir),
    {ok,?TESTMOD,Bin} = compile:forms(dummy_ast(), []),
    ok = file:write_file(filename:join(Dir, ?TESTMODOBJ), Bin),

    {ok, Peer, Node} = ?CT_PEER(["-pa",Dir]),
    {_,_,_} = erpc:call(Node, code, get_object_code, [?TESTMOD]),
    true = erpc:call(Node, code, del_path, [Dir]),
    error = erpc:call(Node, code, get_object_code, [?TESTMOD]),
    ok.
```

For the fix, I proposed above to clear the load paths inside `erl_prim_loader`, but that would be a breaking change. Another option is to store the `-loader` inside the `code_server` and only fallback to `erl_prim_loader` if using the `inet` loader.
